### PR TITLE
Implementing the previous isolation guidance

### PR DESCRIPTION
**Pull request by erik-rosenstrom**
This module implements the previous isolation guidance. This can be thought of a decision tree shown in the image below. The logic of the policy is handled my two separate methods `isolation` and `handle_symptom_resolution`
- `isolation`: set the individual to begin isolating and schedules a follow up test if necessary
- `handle_symptom_resolution`: at the time the symptoms of an individual end, depending on the individuals last test result and their symptom category, determines how long they should remain in isolation, and if they should follow post isolation pre caution. 

There are five tests for the policy. The first 3 tests infect a single person and observe their progression through the policy decision tree. It is likely that test 3 could be extended to capture tests 1 and 2, but for now they are left separate.
1. Test sensitivity is set to 1, and we observe that all individuals end up in leaves 4 or 5
2. Test sensitivity is set to 0, and we observe that all individuals end in up in leaf 1
3. Test sensitivity is set to 0.5 and we only observe individuals that end up in leaves 2 and 3. If individuals are headed toward leaves 1, 4 or 5, the simulation is killed. 
4. Test that the correct proportion of people start the policy
5. Test that policy input throws errors correctly.

### Note 
There is a small increase added to the sampled symptom duration to avoid it being approximately 0.

image

### CODE REVIEW

File: src/policies/previous_guidance.rs [hunk 1]
```diff
@@ -0,0 +1,1010 @@
...
+            let itinerary = vec![
+                ItineraryEntry::new(SettingId::new(Home, 0), 1.0),
+                ItineraryEntry::new(SettingId::new(CensusTract, 0), 1.0),
+                ItineraryEntry::new(SettingId::new(Workplace, 0), 1.0),
+            ];
+            context.add_itinerary(p1, itinerary).unwrap();
+            crate::symptom_progression::init(&mut context).unwrap();
+            super::init(&mut context).unwrap();
+
+            let mild_policy_ran_flag = Rc::new(RefCell::new(false));
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:403]: I think you do need it because you are modifying it within a subscription to event closure.

File: src/policies/updated_guidance.rs [hunk 2]
```diff
@@ -194,6 +194,10 @@ mod test {
             initial_recovered: 0.35,
             proportion_asymptomatic,
             max_time: 100.0,
+            infectiousness_rate_fn: RateFnType::EmpiricalFromFile {
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/updated_guidance.rs:191]: This was a change when I was investigating the short symptom durations. I can revert it back

File: src/policies/previous_guidance.rs [hunk 3]
```diff
@@ -0,0 +1,1010 @@
...
+                        let symptom_severity =
+                            context.get_person_property(event.person_id, SymptomSeverity);
+                        context.set_person_property(
+                            event.person_id,
+                            IsolationEndTime,
+                            context.get_current_time(),
+                        );
+                        match symptom_severity {
+                            SymptomValue::Category1 | SymptomValue::Category2 => {
+                                assert_almost_eq!(
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:463]: I think it should be the max for the same reasons outlined a few comments up.

File: src/policies/previous_guidance.rs [hunk 4]
```diff
@@ -0,0 +1,1010 @@
...
+            define_person_property_with_default!(SymptomStartTime, f64, 0.0);
+            define_person_property_with_default!(SymptomEndTime, f64, 0.0);
+            define_person_property_with_default!(
+                SymptomSeverity,
+                SymptomValue,
+                SymptomValue::Presymptomatic
+            );
+            define_person_property_with_default!(IsolationEndTime, f64, 0.0);
+
+            context.subscribe_to_event::<PersonPropertyChangeEvent<PresentingWithSymptoms>>(
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:419]: I have made changes according to this comment. I just want to note that some person properties are required for the tests that are not required to implement the policy. Those remain in the test.

File: src/symptom_progression.rs [hunk 5]
```diff
@@ -203,7 +202,7 @@ fn schedule_recovery(data: &SymptomData, context: &Context) -> (Option<SymptomVa
         sample
     });
     // Schedule the person to recover from their symptoms (`Symptoms` = `None`) at the given time
-    (None, time)
+    (None, time + 0.000_001)
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/symptom_progression.rs:211]: For now I am going to remove this new file. The issues surrounding these small times have been identified as being related to the ordering of plans in the queue. This is addressed in the test by the same assumption/logic that if an individual's symptoms improve before they begin isolating they will not follow the policy. This is true even in the extreme case of an individual whose symptoms improve instantaneously in combination with a delay to begin the policy of 0.

File: src/policies/previous_guidance.rs [hunk 6]
```diff
@@ -0,0 +1,1010 @@
...
+                            context.administer_test(person_id, intervention_policy_parameters);
+                            trace!("Person {person_id} was tested");
+                        }
+                    },
+                );
+            }
+        }
+    }
+
+    fn post_isolation(
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:92]: I have made changes in both the previous and updated guidance to improve the naming.

**Review by KOVALW — COMMENTED** [review]
LGTM! Some small nits for interpretability

File: src/policies/previous_guidance.rs [hunk 7]
```diff
@@ -0,0 +1,1033 @@
...
+        if self
+            .get_person_property(person_id, LastTestResult)
+            .unwrap_or(false)
+        {
+            if let Some(symptom_record) = self.get_person_property(person_id, SymptomRecord) {
+                let isolation_start_time = symptom_record.symptom_start
+                    + intervention_policy_parameters.isolation_delay_period;
+                let minimum_isolation_time = if symptom_record.severe {
+                    intervention_policy_parameters.moderate_symptom_isolation_duration
+                        + isolation_start_time
```
Comment (eqmooring) [inline src/policies/previous_guidance.rs:158]:
```suggestion
                        + isolation_start_time - intervention_policy_parameters.isolation_delay_period
```
This is probably not the most natural / best way to do this, but I am making this suggestion because I want to flag that `moderate_symptom_isolation_duration` and `mild_symptom_isolation_duration` are defined in terms of time since symptom onset. (I.e., if a person delays entering isolation, there's no biological or public health reason for them to have to stay in isolation a longer amount equivalent to the delay).

File: src/policies/previous_guidance.rs [hunk 8]
```diff
@@ -0,0 +1,1033 @@
...
+                } else {
+                    intervention_policy_parameters.mild_symptom_isolation_duration
+                        + isolation_start_time
```
Comment (eqmooring) [inline src/policies/previous_guidance.rs:161]: See comment above.

File: src/policies/previous_guidance.rs [hunk 9]
```diff
@@ -0,0 +1,1033 @@
...
+                    if !symptom_record.severe {
+                        context.make_post_isolation_masking_plan(
+                            person_id,
+                            intervention_policy_parameters,
+                        );
+                    }
+                });
+            }
+        } else {
+            self.modify_isolation_status(person_id, false).unwrap();
```
Comment (eqmooring) [inline src/policies/previous_guidance.rs:176]: Is there a need for a line here canceling any scheduled test is symptoms improve after a first (negative) test and before it's time for the second test? Is this happening implicitly somewhere in the logic?

Is there a need to cancel any "first" tests? The first test is tied to entering isolation (right?) so perhaps this is handled elsewhere.

File: src/policies/previous_guidance.rs [hunk 10]
```diff
@@ -0,0 +1,1033 @@
...
+    delay_to_retest: f64,
+    policy_adherence: f64,
+    isolation_delay_period: f64,
+    test_sensitivity: f64,
+}
+
+trait ContextIsolationGuidanceInternalExt:
+    PluginContext + ContextRandomExt + ContextPeopleExt + ContextSettingExt
+{
+    fn modify_isolation_status(
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:37]: `IsolatingStatus` is the name of the person property in the model and `isolation_status` is the value we are setting for the person property. 

I do like adding itinerary into the method name. How about something like `modify_itinerary_and_isolation_status`?

File: src/policies/previous_guidance.rs [hunk 11]
```diff
@@ -0,0 +1,1033 @@
...
+                    retest = true;
+                }
+                self.set_person_property(person_id, LastTestResult, Some(false));
+            }
+        } else {
+            if last_test_result.is_none() {
+                retest = true;
+            }
+            self.set_person_property(person_id, LastTestResult, Some(false));
+        }
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:136]: Given that individuals start the symptom progression process once an individual enters the infectious compartment (worth noting this implementation is specific to empirical rate functions), no susceptible individuals will ever be tested.

Yes I think it is important to get an accurate count of the number of tests to inform cost effectiveness analysis later down the line.

File: src/policies/previous_guidance.rs (hunk 9 above)
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:176]: Before `context.administer_test_and_schedule_any_followup` is called (even when called through a plan for the future) there is an if statement ensuring that the function is called only if the person is symptomatic. So, while an excess plan may be firing, the testing method will only be called when necessary.

File: src/policies/previous_guidance.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (eqmooring) [reply src/policies/previous_guidance.rs:37]: I think that's fine. I see now that there is a sense in which this function "modifies" isolation status (line 42).

File: src/policies/previous_guidance.rs (hunk 9 above)
Comment (eqmooring) [reply src/policies/previous_guidance.rs:176]: Ah, upon looking back, I see how symptom improvement is checked in lines 59 and 99.

File: src/policies/updated_guidance.rs [hunk 12]
```diff
@@ -271,39 +274,25 @@ mod test {
...
-
         context.subscribe_to_event::<PersonPropertyChangeEvent<IsolatingStatus>>(
             move |context, event| {
                 if event.current {
                     assert_almost_eq!(
-                        context.get_person_property(event.person_id, SymptomStartTime)
+                        context
+                            .get_person_property(event.person_id, SymptomRecord)
+                            .unwrap()
+                            .symptom_start
```
Comment (eqmooring) [inline src/policies/updated_guidance.rs:284]: I don't really know, but looking at code in lines e.g., 312 and 322 and perhaps naively thinking about pattern matching, is an `.unwrap()` needed here?

**Review by eqmooring — APPROVED** [review]
I think this is looking really good! Let me know if you have any questions, but I don't think I need to re-review.

File: src/policies/previous_guidance.rs [hunk 13]
```diff
@@ -0,0 +1,1034 @@
...
+                .unwrap()
+                .symptom_end
+            {
+                long_delay_flag = symptom_duration
+                    - context
+                        .get_person_property(p1, SymptomRecord)
+                        .unwrap()
+                        .symptom_start
+                    <= isolation_delay_period;
+            }
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/previous_guidance.rs:877]: The difference arises because in this test we shut down if individuals do not progress down the policy decison tree that we are interested in. In this case the simulation ends potentially before individuals are assigned a symptom end time, meaning the value is still `None`. We therefore have to treat it slightly differently when evaluating it.

File: src/policies/updated_guidance.rs [hunk 14]
```diff
@@ -271,39 +274,25 @@ mod test {
         let policy_ran_flag = Rc::new(RefCell::new(false));
         let policy_ran_flag_clone = Rc::clone(&policy_ran_flag);
 
-        define_person_property_with_default!(SymptomStartTime, f64, 0.0);
-        define_person_property_with_default!(SymptomEndTime, f64, 0.0);
-
-        context.subscribe_to_event::<PersonPropertyChangeEvent<PresentingWithSymptoms>>(
```
[the remark this answers is not in the document]
Comment (erik-rosenstrom) [reply src/policies/updated_guidance.rs:277]: Its really just to have a cleaner approach using symptom records. It provides better modularity with other portions of the code.
